internal/storage/ipfs.go

```diff
@@ -0,0 +1,132 @@
+package storage
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"mime/multipart"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	"github.com/covalenthq/bsp-agent/internal/config"
+	"github.com/ipfs/go-cid"
+	log "github.com/sirupsen/logrus"
+	"github.com/ybbus/httpretry"
+)
+
+type ipfsStore struct {
+	baseURL    string // url of ipfs-pinner node
+	httpClient *http.Client
+}
+
+func newIpfsStore(conf *config.StorageConfig) (*ipfsStore, error) {
+	store := ipfsStore{}
+	log.Infof("ipfs store pointint at: %s", conf.IpfsPinnerServer)
+	if _, err := url.Parse(conf.IpfsPinnerServer); err != nil {
+		return nil, fmt.Errorf("url parse error: %w", err)
+	}
+	store.baseURL = conf.IpfsPinnerServer
+	store.httpClient = newHTTPClient(nil)
+
+	return &store, nil
+}
+
+// Upload uploads given content to ipfs
+func (store *ipfsStore) Upload(contents []byte) (cid.Cid, error) {
+	return store.fetchCid(contents, true)
+}
+
+// Cid finds the cid of given content
+func (store *ipfsStore) Cid(contents []byte) (cid.Cid, error) {
+	return store.fetchCid(contents, false)
+}
```

Comment:
Why have you decided to go down the route of calling the same function? in case files already uploaded?